File: leetcode/1493. 删掉一个元素以后全为 1 的最长子数组.cpp
#include <vector>
using namespace std;
class Solution {
public:
    int longestSubarray(vector<int>& nums) {
        vector<pair<int, int>> dp;
        int k = 0;
        int maxLen = 0;
        while (k < nums.size()) {
            if (nums[k] == 1) {
                int count = 0;
                while (k < nums.size() && nums[k] == 1) {
                    count++;
                    k++;
                }
                dp.push_back({k - 1, count});
                maxLen = max(maxLen, count);
                continue;
            }
            k++;
        }
        if (maxLen == nums.size()) {
            return maxLen - 1;
        }
        if (dp.size() == 0) {
            return maxLen;
        }
        for (int i = 0; i < dp.size() - 1; i++) {
            if (dp[i + 1].first - dp[i + 1].second - dp[i].first == 1) {
                maxLen = max(maxLen, dp[i].second + dp[i + 1].second);
            }
        }
        return maxLen;
    }
};

File: leetcode/1493. 删掉一个元素以后全为 1 的最长子数组.rs
impl Solution {
    pub fn longest_subarray(nums: Vec<i32>) -> i32 {
        let mut dp: Vec<(i32, i32)> = Vec::new();
        let mut k = 0;
        let mut max_len = 0;
        while k < nums.len() {
            if nums[k] == 1 {
                let mut count: i32 = 0;
                while k < nums.len() && nums[k] == 1 {
                    count += 1;
                    k += 1;
                }
                dp.push(((k - 1) as i32, count));
                max_len = max_len.max(count);
            } else {
                k += 1;
            }
        }
        if max_len == nums.len() as i32 {
            return max_len - 1;
        }
        if dp.len() == 0 {
            return 0;
        }
        for i in 0..dp.len() - 1 {
            if dp[i + 1].0 - dp[i + 1].1 - dp[i].0 == 1 {
                max_len = max_len.max(dp[i].1 + dp[i + 1].1);
            }
        }
        max_len
    }
}